pkg/datapath/linux/node.go

```diff
@@ -1368,6 +1368,50 @@ func (n *linuxNodeHandler) NodeNeighborRefresh(ctx context.Context, nodeToRefres
 	}
 }
 
+// NodeCleanNeighbors cleans all neighbor entries for the direct routing device
+// and the encrypt interface. It should be used when the agent changes the
+// state from `n.enableNeighDiscovery = true` to
+// `n.enableNeighDiscovery = false`.
+func (n *linuxNodeHandler) NodeCleanNeighbors() {
+	linksToClean := []string{option.Config.DirectRoutingDevice}
+	if option.Config.EncryptInterface != option.Config.DirectRoutingDevice {
+		linksToClean = append(linksToClean, option.Config.EncryptInterface)
+	}
+
+	for _, linkName := range linksToClean {
+		l, err := netlink.LinkByName(linkName)
+		if err != nil {
+			if _, ok := err.(netlink.LinkNotFoundError); !ok {
+				log.WithError(err).WithFields(logrus.Fields{
+					logfields.Device: linkName,
+				}).Error("Unable to GC old neighbor entries for network device")
+			}
+			continue
+		}
+
+		neighList, err := netlink.NeighListExecute(netlink.Ndmsg{
+			Family: netlink.FAMILY_V4,
+			Index:  uint32(l.Attrs().Index),
+			State:  netlink.NUD_PERMANENT,
+		})
+		if err != nil {
+			log.WithError(err).WithFields(logrus.Fields{
+				logfields.Device:    linkName,
+				logfields.LinkIndex: l.Attrs().Index,
+			}).Error("Unable to GC old neighbor entries for network device")
+			continue
+		}
+		for _, neigh := range neighList {
+			err := netlink.NeighDel(&neigh)
```

Comment:
Is there any chance this will lead to connectivity blips during an upgrade?

Or put another way, if we delete these, will the datapath automatically handle this during the next request or does it require userspace to separately run some logic to evaluate the neighbors?

Reply:
the kernel should be able to do this automatically and it will simply do L2 broadcast requesting for the address of this that particular IP address.